sdk_refactor_design_proposal.md

```diff
@@ -0,0 +1,390 @@
+
+   Because of all these possible changes, we need to encapsulate them so that when changes arise, they are contained, without effecting the system.
+
+For these reasons, I propose we extract the configuration and credential information to their own objects to be created by the user before the Client is called. In those classes, all the collection and validations will be performed. As a result, the Client constructor will not be forced to be in charge of handle multiple responsibilities, the user will have to explicitly define the and the areas of volatility will be contained.
+
+We can handle the separation in one of the following ways:
+
+|      | Solution                                        | Description                                                  | Pros                                                         | Cons                                                         |
+| ---- | ----------------------------------------------- | ------------------------------------------------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| 1    | Keep as is                                      |                                                              | - Doesn't break current user apps<br />- Seems to be the direction the industry is going | - 92 lines and 9 params so far makes it unweidy<br />- Constructor is overloaded, handling multiple responsibilities<br />- Doesn't take advantage of OOP |
+| 2    | Config/Creds objects for *each* supported flow  | ConfigurationFromFile()  - .conjurrc<br />ConfigurationFromEnv() - ENV vars<br />CredentialsFromFile() - netrc<br />CredentialsFromKeystore() - keystore | - Explicit, the flow the user chooses is very clear<br />- Security is up to user's the discretion | - Conflicting experience with CLI. The CLI performs a discovery on Config/Creds retrieval type<br />- Requires changes when a user wants to choose another flow |
+| 3    | Generic Config/Creds object                     | Configuration()<br />Credentials()<br />and we will discover where the connection details live based on environment's context | - Easy maintainability. No change in user's code as we build new ways to fetch config and creds<br />- Aligns with CLI UX of Config/Cred discovery | - Implicit and the flow we choose is not clear               |
+|      | A single Config/Cred object with defining param | Configuration(action=file) - .conjurrc<br />Credential(action=keystore) | - Explicit<br />- Minimal code changes should the user choose a different flow<br />- Strong drive in Python to be as explicit as possible | - Requires changes in code if user wants to adopt another retrieval method<br />- Conflicting experience with CLI. The CLI preforms a discovery on Config/Creds retrieval type |
+
+**Decision:** Option #4. 
```

Comment:
Why option 4 and not 2 . I agree with not choosing 1 and 3?

Reply:
After giving at another look I agree 2 is better. we can expose credentialProviders as part of our module so the user can create CredentialData object from file or keystore. but the Client should get as a parameter CredentialsData object

Reply:
@mbenita-Cyberark - can you explain why 2 is better? we can expose a provider also in no. 4.
IMHO we don't need this extensibility point as I can't see a flow where you want the consumer of the config control the internal of where to get it from. It is a leakage of the DAL into a higher layer IMO

Reply:
@mbenita-Cyberark I intended the user to be able to create a CredentialObject/ConfigurationObject regardless. See under "Proposed flow", the following sentence:

> For both Configuration and Credential flows, we will still support the ability for users to build these objects themselves

Reply:
@Tovli the keyStoreCredentialProvider can do CRUD operations on the keystore this is powerful, cross-platform and safer tool to use since it lowers the chance the user will write the credentials hard-coded in his/her app. But maybe we don't to expose this since it's another interface we need to maintain.
So I'm in favor to simply expose the user a `Credentials` and `Configurations` classes and let the user construct them in his own way.
Please note that python doesn't support constructor overloading. so supporting both `Credential(action=keystore) ` and `Credential(....) ` api will be pretty messy since it requires *`*kwargs` that we wanted to avoid.